prefect_dbt/cloud/clients.py

```diff
@@ -72,6 +72,64 @@ async def get_run(self, run_id: int) -> Response:
 
         return response
 
+    async def list_run_artifacts(
+        self, run_id: int, step: Optional[int] = None
+    ) -> Response:
+        """
+        Sends a request to the [list run artifacts endpoint](https://docs.getdbt.com/dbt-cloud/api-v2#tag/Runs/operation/listArtifactsByRunId)
+        to fetch a list of paths of artifacts generated for a completed run.
+
+        Args:
+            run_id: The ID of the run to list run artifacts for.
+            step: The index of the step in the run to query for artifacts. The
+                first step in the run has the index 1. If the step parameter is
+                omitted, then this method will return the artifacts compiled
+                for the last step in the run.
+
+        Returns:
+            The response from the dbt Cloud administrative API.
+        """  # noqa
+        params = dict()
+        if step:
+            params["step"] = step
+        response = await self._admin_client.get(
+            f"/runs/{run_id}/artifacts/", params=params
+        )
+
+        response.raise_for_status()
+
+        return response
+
+    async def get_run_artifact(
+        self, run_id: int, path: str, step: Optional[int] = None
+    ) -> Response:
+        """
+        Sends a request to the [get run artifact endpoint](https://docs.getdbt.com/dbt-cloud/api-v2#tag/Runs/operation/getArtifactsByRunId)
+        to fetch an artifact generated for a completed run.
+
+        Args:
+            run_id: The ID of the run to list run artifacts for.
+            path: The relative path to the run artifact (e.g. manifest.json, catalog.json,
+                run_results.json)
+            step: The index of the step in the run to query for artifacts. The
+                first step in the run has the index 1. If the step parameter is
+                omitted, then this method will return the artifacts compiled
+                for the last step in the run.
+
+        Returns:
+            The response from the dbt Cloud administrative API.
+        """  # noqa
+        params = dict()
+        if step:
+            params["step"] = step
+        response = await self._admin_client.get(
+            f"/runs/{run_id}/artifacts/{path}", params=params
+        )
+
+        response.raise_for_status()
+
+        return response
```

Comment:
In addition to returning the response, can you write to a file and/or store json artifacts as dicts and sql artifacts as strings in memory? That's what dbt developers actually need to pass through into subsequent steps in a pipeline to rerun from point of failure.

Reference implementation: [here](https://github.com/apache/airflow/blob/main/airflow/providers/dbt/cloud/operators/dbt.py#L211)

Example Pipeline Use Case: [here](https://github.com/sungchun12/airflow-dbt-cloud/blob/5e58b9941e61fea6592dd280246a2c3d8ca391e7/archive/dbt_cloud_smart_reruns.py#L72)

Reply:
Example Code

```python
        with open(path, "w") as file:
            if path.endswith(".json"):
                json.dump(response.json(), file)
            else:
                file.write(response.text)
```

Reply:
You can ignore this comment because I verified it's already stored in memory as a dictionary with `artifact.result()`

Thanks!

Reply:
I added an example to the `get_dbt_cloud_run_artifact` task showing how to write the output of that task to a file to guide users that want to do that are part of their flow.